File: python3/implement_stack_using_queues.py
from collections import deque

"""
Space     : O(n)
Push Time : O(n)
Pop Time  : O(1)
"""

class MyStack:

    def __init__(self):
        self.queue = deque()

    def push(self, x: int) -> None:
        temp = deque([x])
        temp.extend(self.queue)
        self.queue = temp

    def pop(self) -> int:
        return self.queue.popleft()

    def top(self) -> int:
        return self.queue[0]

    def empty(self) -> bool:
        return len(self.queue) == 0
        

File: golang/implement_stack_using_queues.go
package main

type MyStack struct {
	Q []int
}

func MyStackConstructor() MyStack {
	return MyStack{
		Q: []int{},
	}
}

func (this *MyStack) Push(x int) {
	this.Q = append([]int{x}, this.Q...)
}

func (this *MyStack) Pop() int {
	n := this.Q[0]
	this.Q = this.Q[1:]
	return n
}

func (this *MyStack) Top() int {
	return this.Q[0]
}

func (this *MyStack) Empty() bool {
	return len(this.Q) == 0
}

/**
 * Your MyStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * param_2 := obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.Empty();
 */
